### ISSUE TITLE
Speeding Up Evaluation of Padded Time Series Using Events

### ISSUE THREAD

**Issue by jacob-hjortlund** [patrick-kidger/diffrax#602]
I am currently training an online NCDE on a set of time series that are irregularly sampled and have varying lengths. To batch these time series, I pad them with the last (t_{N_i}, x_{N_i}) pair such that each time series has the same number of observations N_max, where N_i and N_max are the number of observations in the i'th and most observed time series, respectively.

I need to evaluate the NCDE at each observation in a time series due to the online nature of the model. Since in most cases N_i << N_max, this results in a large amount of wasted calculation. My thinking was to use an Event to stop the integration early for each time series. I've implemented a MWE below, where I've replaced the NCDE with an ODE.

`
import jax
jax.config.update("jax_enable_x64", True)

import diffrax
import numpy as np
from time import time

def f(t, y, args):
    return -y

N = 50
step = 0.2
cols = np.arange(N+1) * step
rows = np.arange(N+1)[:, None]
ts = np.clip(cols, 0, rows)[1:]

@jax.jit
def solve_ode_with_event(ts, idx_max):

    def cond_fn(t, y, args, **kwargs):

        return t >= jax.lax.dynamic_index_in_dim(ts, idx_max, keepdims=False)

    term = diffrax.ODETerm(f)
    solver = diffrax.Tsit5()
    saveat = diffrax.SaveAt(ts=ts)
    event = diffrax.Event(cond_fn)

    y0 = 1
    solution = diffrax.diffeqsolve(
        term,
        solver,
        t0=ts[0],
        t1=ts[-1],
        dt0=0.1,
        y0=y0,
        saveat=saveat,
        event=event
    )

    return solution

@jax.jit
def solve_ode(ts):

    term = diffrax.ODETerm(f)
    solver = diffrax.Tsit5()
    saveat = diffrax.SaveAt(ts=ts)
    stepsize_controller = diffrax.PIDController(rtol=1e-3, atol=1e-6)

    y0 = 1
    solution = diffrax.diffeqsolve(
        term,
        solver,
        t0=ts[0],
        t1=ts[-1],
        dt0=0.1,
        y0=y0,
        saveat=saveat,
        stepsize_controller=stepsize_controller
    )

    return solution

_ = solve_ode(ts[0])
_ = solve_ode_with_event(ts[0], 5)

n_runs = int(1e4)

t_odes = np.zeros(N)
for i in range(N):
    print(i)
    t0 = time()
    for j in range(n_runs):
        _ = solve_ode(ts[i])
    t_ode = (time() - t0) / n_runs
    t_odes[i] = t_ode

t_events = np.zeros(N)
for i in range(N):
    t0 = time()
    for j in range(n_runs):
        _ = solve_ode_with_event(ts[i], i+1)
    t_event = (time() - t0) / n_runs
    t_events[i-1] = t_event

speed_ups = t_odes / t_events
`

Running this example, I find that the event-based approach is slower across the board compared to simply just running the diffeqsolve over the padded time series and does not seem to vary significantly with the choice of time to stop integration at.

As far as I understand, in the case where an event is detected then the remaining steps of the solver are calculated using the value of the ode at t_event to ensure static array shapes, after which these steps are masked out. I'd think the observed slow down when using Events is then simply due to the additional overhead of evaluating events at each step.

Is this understanding correct? Or would we be able to get speed ups using this approach for more costly ODE evaluations? This relates somewhat to #601 .

**Comment by patrick-kidger** [comment]
I think your understanding is correct. You've introduced an overhead by checking for events. But you still have to take just as many steps as before to integrate all the batch elements which haven't triggered the event (or end of solve) yet.

**Comment by jacob-hjortlund** [comment]
Gotcha, thanks for confirming my thinking. Would it maybe be possible to speed this up by removing the restriction that `saveat.ts > t1` in `diffeqsolve`? https://github.com/patrick-kidger/diffrax/blob/034ead8fbb53878bd140118a8e7024eb5968db50/diffrax/_integrate.py#L1195-L1206

This would also require modifying the outer `cond_fun` in `loop` to only have one branch with `out = state.tprev < t1`, which ofc isn't ideal. https://github.com/patrick-kidger/diffrax/blob/034ead8fbb53878bd140118a8e7024eb5968db50/diffrax/_integrate.py#L331-L339

Additionally, I've added the condition `ts[_save_state.saveat_ts_index] <= t1` to the `cond_fun` of `save_ts_impl` to similarly reduce the number of steps taken in the inner while loop for save state evaluation. https://github.com/patrick-kidger/diffrax/blob/034ead8fbb53878bd140118a8e7024eb5968db50/diffrax/_integrate.py#L438-L444

As far as I've understood, memory for saving at specific times is allocated in `diffeqsovle`, and values are filled in by `loop`. If we make the modifications above then the outer while loop should terminate early, leading to fewer solver steps and saving on compute. In my previous MWE I had forgotten to add `jump_ts=ts` to the step-size controller, which is necessary due to the use of rectilinear interpolation. With this added, then any reduction in number of steps should hopefully lead to performance increases. 

I've modified my example from previously to test this with the mentioned changes in `_integrate.py`. 

```
import jax
jax.config.update("jax_enable_x64", True)

import time
import diffrax
import numpy as np
import equinox as eqx
from tqdm import tqdm

def f(t, y, args):
    return -y

N = 50
step = 0.2
cols = np.arange(N+1) * step
rows = np.arange(N+1)[:, None]
ts = np.clip(cols, 0, rows)[1:]

@jax.jit
def solve_ode(ts):

    term = diffrax.ODETerm(f)
    solver = diffrax.Tsit5()
    saveat = diffrax.SaveAt(ts=ts)
    stepsize_controller = diffrax.PIDController(rtol=1e-3, atol=1e-6, jump_ts=ts)

    y0 = 1
    solution = diffrax.diffeqsolve(
        term,
        solver,
        t0=ts[0],
        t1=ts[-1],
        dt0=None,
        y0=y0,
        saveat=saveat,
        stepsize_controller=stepsize_controller
    )

    return solution

@jax.jit
def solve_ode_with_low_t1(ts, idx_max):

    term = diffrax.ODETerm(f)
    solver = diffrax.Tsit5()
    saveat = diffrax.SaveAt(ts=ts)
    stepsize_controller = diffrax.PIDController(rtol=1e-3, atol=1e-6, jump_ts=ts)

    y0 = 1
    solution = diffrax.diffeqsolve(
        term,
        solver,
        t0=ts[0],
        t1=ts[idx_max],
        dt0=None,
        y0=y0,
        saveat=saveat,
        stepsize_controller=stepsize_controller
    )

    return solution


_ = solve_ode(ts[0])
_ = solve_ode_with_low_t1(ts[0], 5)

n_runs = int(1e4)

t_odes = np.zeros(N)
nsteps_odes = np.zeros(N)
for i in tqdm(range(N)):
    t0 = time.time()
    for j in range(n_runs):
        sol = solve_ode(ts[i])
    t_ode = (time.time() - t0) / n_runs
    t_odes[i] = t_ode
    nsteps_odes[i] = sol.stats['num_steps']

t_t1s = np.zeros(N)
nsteps_t1s = np.zeros(N)
for i in tqdm(range(N)):
    t0 = time.time()
    for j in range(n_runs):
        sol = solve_ode_with_low_t1(ts[i], i+1)
    t_t1 = (time.time() - t0) / n_runs
    t_t1s[i-1] = t_t1
    nsteps_t1s[i] = sol.stats['num_steps']

speed_up = t_odes / t_t1s
step_ratio = nsteps_odes / nsteps_t1s
```

Results are shown in the figure below. As expected the number of steps taken is reduced when max_idx < len(ts) and converges to the reference value as max_idx increases. Oddly enough this method is still slower than the reference approach, which I haven't found an explanation for yet. Adding the new conditional statements to the outer and inner while loops should have the same overhead per evaluation for both methods, so I'm not really sure where this additional overhead would be coming from.

![Image](https://github.com/user-attachments/assets/73fe50de-fbcd-40a5-9e8a-f291467f3e6e)

**Comment by patrick-kidger** [comment]
> Would it maybe be possible to speed this up by removing the restriction that saveat.ts > t1 in diffeqsolve?

I don't see how that would help? You can't ever fill in a value that is past the terminal time.

> If we make the modifications above then the outer while loop should terminate early, leading to fewer solver steps and saving on compute.

I also don't see why this would be true I'm afraid!

**Comment by jacob-hjortlund** [comment]
> I don't see how that would help? You can't ever fill in a value that is past the terminal time.

That's exactly what I'm after. The ts values passed to SaveAt contain a large number of padding values to ensure a certain array shape to enable batching, but in actuality we only want to save values up to and including t1.

> I also don't see why this would be true I'm afraid!

My understanding is that the solver will step to every value in ts if we pass jump_ts=ts to the stepsize_controller. If we make the modifications above such that saveat.ts > t1, then we can by definition not step to all values in ts before integration completes. In this way then the number of steps would be reduced.

I agree that my statement would not be true if we do not pass either step_ts=ts or jump_ts=ts to the stepsize controller.

**Comment by jacob-hjortlund** [comment]
Alright, so I thought a little further about this, and realized a couple of mistakes in my two previous posts.

You're completely right that we don't need to modify the outer while loop, I had missed that the branching was only for instances of `StepTo`. I also realized that I was not doing an apples-to-apples comparison since `solve_ode_with_low_t1` had additional overhead due to indexing into ts. With this said, then the following modifications were made to `_integrate.py`:

Allow saveat.ts > t1,

```
def _check_subsaveat_ts(ts):
    ts = eqxi.error_if(
        ts,
        ts[1:] < ts[:-1],
        "saveat.ts must be increasing or decreasing.",
    )
    #ts = eqxi.error_if(
    #    ts,
    #    (ts > t1) | (ts < t0),
    #    "saveat.ts must lie between t0 and t1.",
    #)
    return ts
```

and update the conditional of the inner while loop,

```
def save_ts_impl(ts, fn, save_state: SaveState) -> SaveState:
    def _cond_fun(_save_state):
        return (
            keep_step
            & (ts[_save_state.saveat_ts_index] <= state.tnext)
            & (_save_state.saveat_ts_index < jnp.sum(ts <= t1))
            #& (_save_state.saveat_ts_index < len(ts))
            #& (ts[_save_state.saveat_ts_index] <= t1)
        )
```

The updated MWE:

```
import jax
jax.config.update("jax_enable_x64", True)

import time
import diffrax
import numpy as np
import equinox as eqx
from tqdm import tqdm

def f(t, y, args):
    return -y

N = 50
step = 0.125
cols = np.arange(N+1) * step
rows = np.arange(N+1)[:, None]
ts = np.clip(cols, 0, rows)[1:].astype(float)

@jax.jit
def solve_ode(ts, t0, t1):

    term = diffrax.ODETerm(f)
    solver = diffrax.Tsit5()
    saveat = diffrax.SaveAt(ts=ts)
    stepsize_controller = diffrax.PIDController(rtol=1e-3, atol=1e-6, jump_ts=ts)

    y0 = 1
    solution = diffrax.diffeqsolve(
        term,
        solver,
        t0=t0,
        t1=t1,
        dt0=None,
        y0=y0,
        saveat=saveat,
        stepsize_controller=stepsize_controller
    )

    return solution

@jax.jit
def solve_ode_with_low_t1(ts, t0, t1):

    term = diffrax.ODETerm(f)
    solver = diffrax.Tsit5()
    saveat = diffrax.SaveAt(ts=ts)
    stepsize_controller = diffrax.PIDController(rtol=1e-3, atol=1e-6, jump_ts=ts)

    y0 = 1
    solution = diffrax.diffeqsolve(
        term,
        solver,
        t0=t0,
        t1=t1,
        dt0=None,
        y0=y0,
        saveat=saveat,
        stepsize_controller=stepsize_controller
    )

    return solution


_ = solve_ode(ts[0], ts[0][0], ts[0][-1])
_ = solve_ode_with_low_t1(ts[0], ts[0][0], ts[0][-1])

n_runs = int(1e4)

t_odes = np.zeros(N)
nsteps_odes = np.zeros(N)
for i in tqdm(range(N)):
    t0, t1 = ts[i][0], ts[i][-1]
    time_0 = time.time()
    for j in range(n_runs):
        sol = solve_ode(ts[i], t0, t1)
    t_ode = (time.time() - time_0) / n_runs
    t_odes[i] = t_ode
    nsteps_odes[i] = sol.stats['num_steps']

t_t1s = np.zeros(N)
nsteps_t1s = np.zeros(N)
for i in tqdm(range(N)):
    t0, t1 = ts[i][0], ts[i][i+1]
    time_0 = time.time()
    for j in range(n_runs):
        sol = solve_ode_with_low_t1(ts[i],t0, t1)
    t_t1 = (time.time() - time_0) / n_runs
    t_t1s[i-1] = t_t1
    nsteps_t1s[i] = sol.stats['num_steps']

speed_up = t_odes / t_t1s
step_ratio = nsteps_odes / nsteps_t1s
```

The results are shown in the figure below. We see that for Max Time Index below ~40 that we get speed ups in the range 3-15%. Interestingly enough there is a large speed up when we integrate over the full time series, which I have not been able to find the cause of as of yet.

![Image](https://github.com/user-attachments/assets/6588c39c-d660-4185-bb75-54cf26972a74)
